File: tests/layer1_filters/unit/test_firfilter.py
"""
test_firfilter.py
=================
Python port of:
    layer1_filters/unit/TestFirFilter.m

Covers: firfilter (FIR filter constructor)

API
---
firfilter(name, M)           -> filter dict, DC-centred (fc=0)
firfilter(name, M, fc)       -> FIR filter at normalised fc
firfilter(name, M, fc, norm) -> with explicit normalisation flag

The returned object is expected to be dict-like with keys:
    h        : np.ndarray  impulse response
    offset   : int         = -floor(M/2) for non-causal, 0 for causal
    delay    : int         = 0 by default
    realonly : int         = 0 by default, 1 when 'real' flag given
"""

from __future__ import annotations

import pytest

import numpy as np

# ---------------------------------------------------------------------------
# firfilter – implementation tests
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFirfilterStructImpl:
    """
    MATLAB counterpart: TestFirFilter (struct format section).
    """

    def test_has_required_fields(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 32)
        for field in ("h", "offset", "delay", "realonly"):
            assert field in g, f"Missing field: {field}"

    def test_h_is_numeric_vector(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 32)
        h = np.asarray(g["h"])
        assert h.ndim == 1 and np.issubdtype(h.dtype, np.number)

    def test_default_realonly_is_zero(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 32)
        assert g["realonly"] == 0

    def test_real_flag_sets_realonly(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 32, 0, "real")
        assert g["realonly"] == 1


@pytest.mark.requires_impl
class TestFirfilterLengthImpl:
    """
    MATLAB counterpart: TestFirFilter (impulse response length section).
    """

    @pytest.mark.parametrize("M", [16, 32, 64, 128])
    def test_impulse_response_length(self, needs_impl, M):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", M)
        assert len(np.asarray(g["h"])) == M

    def test_sine_filter_length(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("sine", 48)
        assert len(np.asarray(g["h"])) == 48


@pytest.mark.requires_impl
class TestFirfilterNormImpl:
    """
    MATLAB counterpart: TestFirFilter (energy normalisation section).
    """

    @pytest.mark.parametrize("M", [16, 32, 64])
    def test_energy_norm_default(self, needs_impl, M):
        """Default 'energy': sum(h^2) == 1."""
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", M)
        h = np.asarray(g["h"], dtype=float)
        assert np.sum(h ** 2) == pytest.approx(1.0, abs=1e-10)

    def test_peak_norm(self, needs_impl):
        """'peak': max|h| == 1."""
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 64, 0, "peak")
        h = np.asarray(g["h"])
        assert np.max(np.abs(h)) == pytest.approx(1.0, abs=1e-10)

    def test_scal_multiplies_impulse_response(self, needs_impl):
        """'scal' s: h_scaled == s * h_default."""
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        s  = 2.0
        g1 = firfilter("hann", 32)
        g2 = firfilter("hann", 32, 0, "scal", s)
        np.testing.assert_allclose(np.asarray(g2["h"]), s * np.asarray(g1["h"]),
                                   atol=1e-10 * np.linalg.norm(g1["h"]))


@pytest.mark.requires_impl
class TestFirfilterOffsetImpl:
    """
    MATLAB counterpart: TestFirFilter (offset / delay section).
    """

    def test_non_causal_offset_is_minus_floor_M_over_2(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        M = 32
        g = firfilter("hann", M)
        assert g["offset"] == -M // 2

    def test_causal_offset_is_zero(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 32, 0, "causal")
        assert g["offset"] == 0

    def test_default_delay_is_zero(self, needs_impl):
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 32)
        assert g["delay"] == 0


@pytest.mark.requires_impl
class TestFirfilterTransferFunctionImpl:
    """
    MATLAB counterpart: TestFirFilter (comp_transferfunction section).
    """

    def test_impulse_response_is_real_at_dc(self, needs_impl):
        """DC-centred firfilter: h should be real."""
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        g = firfilter("hann", 32)
        h = np.asarray(g["h"])
        assert np.all(np.isreal(h)), "DC firfilter: impulse response should be real"

    def test_peak_near_centre_freq(self, needs_impl):
        """firfilter at fc=0.25: peak of |H| within 5 bins of expected."""
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        from cool_frames.numpy.filters._filters import comp_transferfunction  # type: ignore
        L  = 1024
        fc = 0.25
        g  = firfilter("hann", 64, fc, "peak")
        H  = comp_transferfunction(g, L)
        peak_bin     = int(np.argmax(np.abs(H)))
        expected_bin = round(fc / 2 * L)
        assert abs(peak_bin - expected_bin) < 5

    def test_filterbank_zero_input(self, needs_impl):
        """Zero input -> zero subband output."""
        from cool_frames.filterbanks import filterbank  # type: ignore
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        Ls = 512
        x  = np.zeros(Ls)
        g  = [firfilter("hann", 32)]
        a  = 4
        c  = filterbank(x, g, a)
        assert np.max(np.abs(c[0])) < 1e-12

    def test_filterbank_output_length(self, needs_impl):
        """filterbank subband has ceil(Ls/a) rows."""
        from cool_frames.filterbanks import filterbank  # type: ignore
        from cool_frames.filters.lowlevel import firfilter  # type: ignore
        Ls  = 512
        rng = np.random.default_rng(42)
        x   = rng.standard_normal(Ls)
        g   = [firfilter("hann", 32)]
        a   = 4
        c   = filterbank(x, g, a)
        assert c[0].shape[0] == int(np.ceil(Ls / a))

File: tests/layer1_filters/unit/TestFirFilter.m
classdef TestFirFilter < matlab.unittest.TestCase
%TESTFIRFILTER  Unit tests for firfilter (FIR filter constructor).
%
%   firfilter(name, M)       — time-domain FIR filter, length M, centred at DC
%   firfilter(name, M, fc)   — FIR filter at centre freq fc (normalized)

    properties
        tol = 1e-10
    end

    methods (TestClassSetup)
        function setupClass(tc)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
        end
    end

    % ── Struct format ────────────────────────────────────────────────────────
    methods (Test)

        function testStructHasRequiredFields(tc)
            g = firfilter('hann', 32);
            tc.verifyTrue(isfield(g, 'h'),        'Missing: h');
            tc.verifyTrue(isfield(g, 'offset'),   'Missing: offset');
            tc.verifyTrue(isfield(g, 'delay'),    'Missing: delay');
            tc.verifyTrue(isfield(g, 'realonly'), 'Missing: realonly');
        end

        function testHIsNumericVector(tc)
            g = firfilter('hann', 32);
            tc.verifyTrue(isnumeric(g.h) && isvector(g.h), ...
                'g.h must be a numeric vector');
        end

        function testScalarInputReturnsStruct(tc)
            g = firfilter('hann', 32);
            tc.verifyTrue(isstruct(g));
        end

        function testDefaultRealonlyIsZero(tc)
            g = firfilter('hann', 32);
            tc.verifyEqual(g.realonly, 0);
        end

        function testRealFlagSetsRealonly(tc)
            g = firfilter('hann', 32, 0, 'real');
            tc.verifyEqual(g.realonly, 1);
        end

    end

    % ── Impulse response length ───────────────────────────────────────────────
    methods (Test)

        function testImpulseResponseLength(tc)
            for M = [16, 32, 64, 128]
                g = firfilter('hann', M);
                tc.verifyEqual(numel(g.h), M, ...
                    sprintf('firfilter: impulse response should have length %d', M));
            end
        end

        function testSineFilterLength(tc)
            M = 48;
            g = firfilter('sine', M);
            tc.verifyEqual(numel(g.h), M);
        end

    end

    % ── Real-valued impulse response ──────────────────────────────────────────
    methods (Test)

        function testImpulseResponseIsReal(tc)
            % firwin windows are real, so h should be real (fc=0)
            g = firfilter('hann', 32);
            tc.verifyTrue(isreal(g.h), ...
                'DC-centred firfilter: impulse response should be real');
        end

    end

    % ── Energy normalization ──────────────────────────────────────────────────
    methods (Test)

        function testEnergyNormDefault(tc)
            % Default 'energy': sum(h.^2) = 1  (Parseval: ||h||^2 = ||H||^2/L)
            g = firfilter('hann', 64);
            tc.verifyEqual(sum(g.h.^2), 1, 'AbsTol', tc.tol, ...
                'firfilter energy norm: sum(h^2) should be 1');
        end

        function testEnergyNormMultipleLengths(tc)
            for M = [16, 32, 64]
                g = firfilter('hann', M);
                tc.verifyEqual(sum(g.h.^2), 1, 'AbsTol', tc.tol, ...
                    sprintf('energy norm failed for M=%d', M));
            end
        end

        function testPeakNorm(tc)
            % 'peak' for firfilter normalises the time-domain impulse response:
            % max|h| = 1.  (This differs from blfilter/freqfilter where 'peak'
            % refers to the frequency-domain magnitude.)
            g = firfilter('hann', 64, 0, 'peak');
            tc.verifyEqual(max(abs(g.h)), 1, 'AbsTol', tc.tol, ...
                'firfilter peak norm: max|h| should be 1');
        end

        function testScalParameter(tc)
            s  = 2.0;
            g1 = firfilter('hann', 32);
            g2 = firfilter('hann', 32, 0, 'scal', s);
            tc.verifyEqual(g2.h, s * g1.h, 'AbsTol', tc.tol * norm(g1.h), ...
                '''scal'' should multiply impulse response');
        end

    end

    % ── Offset / delay ────────────────────────────────────────────────────────
    methods (Test)

        function testNonCausalOffsetIsNegativeHalfLength(tc)
            % Default (non-causal): offset = delay - floor(M/2) = 0 - floor(M/2)
            M = 32;
            g = firfilter('hann', M);
            tc.verifyEqual(g.offset, -floor(M/2));
        end

        function testCausalOffsetIsZero(tc)
            g = firfilter('hann', 32, 0, 'causal');
            tc.verifyEqual(g.offset, 0);
        end

        function testDefaultDelayIsZero(tc)
            g = firfilter('hann', 32);
            tc.verifyEqual(g.delay, 0);
        end

    end

    % ── Centre frequency via comp_transferfunction ────────────────────────────
    methods (Test)

        function testPeakNearCentreFreq(tc)
            L  = 1024;
            fc = 0.25;
            g  = firfilter('hann', 64, fc, 'peak');
            H  = comp_transferfunction(g, L);
            [~, peak_bin] = max(abs(H));
            expected_bin  = round(fc / 2 * L) + 1;
            tc.verifyLessThan(abs(peak_bin - expected_bin), 5, ...
                'firfilter: peak should be near specified fc');
        end

    end

    % ── Filterbank integration ────────────────────────────────────────────────
    methods (Test)

        function testZeroInputYieldsZeroOutput(tc)
            Ls = 512;
            x  = zeros(Ls, 1);
            g  = {firfilter('hann', 32)};
            a  = 4;
            c  = filterbank(x, g, a);
            tc.verifyLessThan(max(abs(c{1})), 1e-12, ...
                'Zero input should give zero subband output');
        end

        function testOutputSubbandLength(tc)
            Ls = 512;
            rng(42);
            x  = randn(Ls, 1);
            g  = {firfilter('hann', 32)};
            a  = 4;
            c  = filterbank(x, g, a);
            tc.verifyEqual(size(c{1}, 1), ceil(Ls / a));
        end

    end

end
